internal/wasm/counts_test.go

```diff
@@ -0,0 +1,298 @@
+package internalwasm
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestModule_ImportFuncCount(t *testing.T) {
+	tests := []struct {
+		name     string
+		input    *Module
+		expected uint32
+	}{
+		{
+			name:  "none",
+			input: &Module{},
+		},
+		{
+			name:  "none with function section",
+			input: &Module{FunctionSection: []Index{0}},
+		},
+		{
+			name:     "one",
+			input:    &Module{ImportSection: []*Import{{Type: ExternTypeFunc}}},
+			expected: 1,
+		},
+		{
+			name:     "one with function section",
+			input:    &Module{ImportSection: []*Import{{Type: ExternTypeFunc}}, FunctionSection: []Index{0}},
+			expected: 1,
+		},
+		{
+			name:     "one with other imports",
+			input:    &Module{ImportSection: []*Import{{Type: ExternTypeFunc}, {Type: ExternTypeMemory}}},
+			expected: 1,
+		},
+		{
+			name:     "two",
+			input:    &Module{ImportSection: []*Import{{Type: ExternTypeFunc}, {Type: ExternTypeFunc}}},
+			expected: 2,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			require.Equal(t, tc.expected, tc.input.ImportFuncCount())
+		})
+	}
+}
+
+// TODO: once we fix up-front validation, this only needs to check zero or one
```

Comment:
ps almost certainly I will delete this and the function for counting memory once there can be only one and it is validated during decode. I'm only putting this in now as there are no such guarantees yet. deleting this later will help show how useful centralized decode logic is!